### PR TITLE
Fix #126. Running jupyterhub under Sirepo.

### LINKED ISSUE #126: Support running jupyterhub under Sirepo

### PR DESCRIPTION
**Pull request by e-carlin**
I believe this covers everything for running jupyterhub at sirepo.com/jupyter.

For dev getting everything running on a remote server using ssh LocalForward required some hacks. There were issues with re-directs dropping the port number and jupyterhub blocking requests due to them being cross origin. In the end I setup a reverse proxy on my mac and edited /etc/hosts to make it all work. My nginx foo isn't the best so maybe there is a simpler way to get it working that doesn't require a reverse proxy.

~/.ss/config on mac:

Host <redacted>
  HostName <redacted>
  Port  <redacted>
  User <redcated>
  LocalForward 8080 sirepo.v4.radia.run:443

nginx on mac:

brew install nginx
cd /usr/local/etc/nginx
sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout cert.key -out crt.pem # for fqdn use sirepo.v4.radia.run
cat <<EOF > /usr/local/etc/nginx/nginx.conf
worker_processes  1;
error_log /usr/local/var/log/nginx/error.log debug;

events {
    worker_connections  1024;
}

http {
    include       mime.types;
    default_type  application/octet-stream;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    sendfile        on;
    keepalive_timeout  65;

    proxy_buffer_size   128k;
    proxy_buffers   4 256k;
    proxy_busy_buffers_size   256k;
    server {
      listen 443 ssl;
      server_name sirepo.v4.radia.run;
      ssl_certificate      cert.pem;
      ssl_certificate_key  cert.key;
      proxy_http_version 1.1;
      proxy_next_upstream error timeout invalid_header http_500 http_502 http_503 http_504;
      proxy_read_timeout 600s;
      proxy_send_timeout 600s;
      proxy_connect_timeout 600s;
      proxy_redirect off;
      proxy_set_header Host $host:$server_port;
      proxy_set_header Proxy "";
      proxy_set_header Via $remote_addr;
      proxy_set_header X-Real-Ip $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_cache_bypass $http_upgrade;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
      location / {
          proxy_pass         https://localhost:8080;
      }
    }
    include servers/*;
}
EOF
brew services start nginx

/etc/hosts on mac:
127.0.0.1 localhost localhost.localdomain sirepo.v4.radia.run

Then visit https://sirepo.v4.radia.run/jupyter in your browser.

### CODE REVIEW

**Comment by e-carlin** [comment]
@robnagler I've addressed all comments. Ready for another review.

**Comment by robnagler** [comment]
Is there a missing commit in container-jupyterhub?

```txt
sirepo_jupyterhub[7968]: ModuleNotFoundError: No module named 'sirepo'
```

I figure it must be a dependency now  in the build.sh, but maybe that wasn't committed?

File: rsconf/component/sirepo_jupyterhub.py [hunk 1]
```diff
@@ -86,6 +86,8 @@ def _env_ok(elem):
         self.install_access(mode='711', owner=z.run_u)
         self.install_directory(self.__run_d)
         self.install_access(mode='700', owner=z.jupyter_run_u)
+        if z.user_d.dirname == '/srv/jupyterhub':
+            self.install_directory(z.user_d.dirname)
         self.install_directory(z.user_d)
```
Comment (e-carlin) [inline rsconf/component/sirepo_jupyterhub.py:91]: Should this line be deleted?

Comment (e-carlin) [reply rsconf/component/sirepo_jupyterhub.py:91]: Disregard. I see what's going on.

**Comment by e-carlin** [comment]
I was using radiasoft/sirepo as the base image in my testing but adding Sirepo as a dep makes considerably more sense. Can you give me access to container-jupyterhub?

**Comment by e-carlin** [comment]
Just adding Sirepo as a dep doesn't work because some imports are missing (ex h5py). How should we resolve? I think using sirepo as the base image is overkill because the image includes all of the codes which jupyterhub doesn't need. OTOH using it as the base image will always means sirepo works...

**Comment by robnagler** [comment]
Does sirepo need h5py if there are no codes imported?

**Comment by robnagler** [comment]
I gave you access

**Comment by e-carlin** [comment]
Thanks.

No [h5py isn't imported in that case](https://github.com/radiasoft/sirepo/blob/3b1657c974f1f64768b4861473f1fc0b02d56113/sirepo/server.py#L37).

**Comment by robnagler** [comment]
I think we should import only those sim_types that are necessary, possibly just jupyterlogin.  The configuration needs to be custom, which we can override inside the build.sh, since it will always be true, not just in rsconf. Some config can't be overriden (secret, for example).

**Comment by e-carlin** [comment]
Ok. You lost me on what build.sh has to do with this? Maybe I wasn't clear above but the image builds fine. It is just when running jupyterhub and it imports sirepo then sirepo imports h5py that the problem arises. So, I think we can set feature config sim_types to be jupyterhublogin in sirepo_jupyterhub's env and all should work fine. Agree?

**Comment by robnagler** [comment]
I see that we don't have a radia-run.sh in container-jupyterhub like we do with sirepo. This means we don't have a hook to set the environment in which jupyter starts up. This has to be in sirepo_juptyerhub, then. It's probably fine, but you'll also need to modify test.sh to pass SIREPO_FEATURE_CONFIG_SIM_TYPES.

**Comment by e-carlin** [comment]
@robnagler I update the sirepo_jupyterhub env in rsconf to set SIREPO_FEATURE_CONFIG_SIM_TYPES. 

I realized container-jupyterhub/container-config/test.sh doesn't need to have the env. It uses a dummy authenticator so Sirepo is never imported.

File: rsconf/package_data/dev/db/000.yml.jinja [hunk 2]
```diff
@@ -64,9 +64,6 @@ default:
           on_calendar: "1 0"
           bash_script: |
             echo doing that thing
-
```
Comment (robnagler) [inline rsconf/package_data/dev/db/000.yml.jinja:67]: needs one blank line after a section that begins with `|`

File: rsconf/package_data/dev/db/000.yml.jinja [hunk 3]
```diff
@@ -240,6 +244,13 @@ channel:
         mail_server: {{ host }}
         mail_username: comsol@{{ host }}
         mail_password: somepass
+      feature_config:
+       proprietary_sim_types:
```
Comment (robnagler) [inline rsconf/package_data/dev/db/000.yml.jinja:248]: indentation incorrect

File: rsconf/package_data/sirepo/nginx.conf.jinja [hunk 4]
```diff
@@ -10,6 +10,21 @@ server {
     root {{ nginx.default_root }};
     ssl_certificate {{ nginx.tls_crt }};
     ssl_certificate_key {{ nginx.tls_key }};
+{% if 'jupyterhub_enabled' in sirepo %}
```
Comment (robnagler) [inline rsconf/package_data/sirepo/nginx.conf.jinja:13]: `{% if sirepo.jupyterhub_enabled %}`

File: rsconf/package_data/sirepo/nginx.conf.jinja [hunk 5]
```diff
@@ -10,6 +10,21 @@ server {
...
+    proxy_http_version 1.1;
```
Comment (robnagler) [inline rsconf/package_data/sirepo/nginx.conf.jinja:14]: [these are in ngnix.conf](https://github.com/radiasoft/rsconf/blob/84db636618318686e142cb0ba13448896919d50f/rsconf/package_data/nginx/global.conf.jinja#L85) so you don't need them here.

File: rsconf/package_data/sirepo/nginx.conf.jinja [hunk 6]
```diff
@@ -18,4 +33,12 @@ server {
         uwsgi_send_timeout 300s;
         uwsgi_pass {{ nginx.backend_host }}:{{ nginx.backend_port }};
     }
+{% if 'jupyterhub_enabled' in sirepo %}
```
Comment (robnagler) [inline rsconf/package_data/sirepo/nginx.conf.jinja:36]: `{% if sirepo.jupyterhub_enabled %}`

File: rsconf/component/sirepo_jupyterhub.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            gen_secret=lambda: db.random_string(length=64, is_hex=True),
+            visibility='channel',
+        )[0]
+        z.admin_users_str = _list_to_str(z.admin_users)
+        self._rsdockerspawner(jc, z)
+
+        self.install_access(mode='711', owner=z.run_u)
+        self.install_directory(self.__run_d)
+        self.install_access(mode='700', owner=z.jupyter_run_u)
+        if z.user_d.dirname == '/srv/jupyterhub':
```
Comment (robnagler) [inline rsconf/component/sirepo_jupyterhub.py:89]: This is annoying. I don't have a solution for it, and we'll need it for non-sirepo jupyter servers until we solve the problem of choosing pools.

**Review by robnagler — CHANGES_REQUESTED** [review]

File: rsconf/package_data/sirepo/nginx.conf.jinja (hunk 4 above)
Comment (e-carlin) [reply rsconf/package_data/sirepo/nginx.conf.jinja:13]: 361965b

File: rsconf/package_data/dev/db/000.yml.jinja (hunk 3 above)
Comment (e-carlin) [reply rsconf/package_data/dev/db/000.yml.jinja]: 361965b

File: rsconf/package_data/dev/db/000.yml.jinja (hunk 2 above)
Comment (e-carlin) [reply rsconf/package_data/dev/db/000.yml.jinja:67]: 361965b

File: rsconf/package_data/sirepo/nginx.conf.jinja (hunk 6 above)
Comment (e-carlin) [reply rsconf/package_data/sirepo/nginx.conf.jinja]: https://github.com/radiasoft/rsconf/pull/128/commits/0090bba595e55686c9e97eb30d38cf762f5612d2

File: rsconf/package_data/sirepo/nginx.conf.jinja (hunk 5 above)
Comment (e-carlin) [reply rsconf/package_data/sirepo/nginx.conf.jinja]: https://github.com/radiasoft/rsconf/pull/128/commits/0090bba595e55686c9e97eb30d38cf762f5612d2

File: rsconf/component/sirepo.py [hunk 8]
```diff
@@ -187,6 +187,9 @@ def _install_dirs_and_files(self):
             self.install_directory(p.join(c))
         self.install_access(mode='400')
         for c in z.feature_config.proprietary_sim_types:
+            # TODO(e-carlin): Need a better way to configure this
+            if c == 'jupyterhublogin':
```
Comment (e-carlin) [inline rsconf/component/sirepo.py:191]: @robnagler how should we ask the question of whether or not a proprietary code has an rpm?

**Comment by e-carlin** [comment]
@robnagler all comments addressed. One [question for you](https://github.com/radiasoft/rsconf/pull/128#discussion_r537196719).

File: rsconf/component/sirepo.py (hunk 8 above)
Comment (robnagler) [reply rsconf/component/sirepo.py:191]: I thought I saw you called proprietary_rpm or something like that. Any interface is fine as it is quite internal.

**Pull request merged**